@awslambda decorator

oavdeev opened the pull request:
This is a PR to implement https://github.com/Netflix/metaflow/issues/358 , stacked on top of #810 

### What works
You can decorate steps with `@awslambda` and they will run on Lambda. It is pretty quick, just a few seconds spin-up time, so the user experience is much better than AWS Batch. Logs are collected and streamed properly.

`@awslambda` respects `@resource`, `@retry` and `@timeout` decorators.

You can also specify a custom docker image like this:
```python
@awslambda(image="1234.dkr.ecr.us-west-2.amazonaws.com/my-image:latest")
def my_step(self):
    ...
```
The image must be pushed to a ECR registry, public images are not supported by AWS Lambda.

### Not supported in this PR
..but could be done as future improvements:

* No integration with Step Functions. It would be nice if one could write mixed Batch+Lambda workflows and deploy them to Step, but this is not implemented here (you'll get a nice error message)
* `@conda` is not supported (you'll get a nice error message). The reason is that Lambdas only get .5 GB of writable disk space. It could probably technically work for smallish conda deps, but even installing Pandas bumps into that limit, so I figured better to disable this altogether. What we could implement instead in a future PR, is to build and push a custom docker image for each `@conda` step, since Lambda supports docker images up to 10GB in size.

### How to give it a try
For now the setup is somewhat involved, hopefully we'll automate away most of this:
1. In your AWS account, create a ECR repo to hold the docker image for Lambda. 
2. Run `Makefile` from https://github.com/Netflix/metaflow-tools/pull/29 to build the image and push it to the repo. 
3. Create a Lambda IAM role that has access to Metaflow S3 bucket.
4. Set `METAFLOW_LAMBDA_ROLE_ARN` env var to lambda role ARN created in step 4
5. Set `METAFLOW_LAMBDA_CONTAINER_IMAGE` to the docker image URI from step (2)

### Other Limitations
There are a couple of limitations with Lambdas that I don't think can be addressed nicely:
1. AWS provides no way to cancel Lambda execution. If we want to support that, we'd have to invent something awkward with sending termination signal to Lambda via SQS or something. For now lambdas will always run to completion.
2. AWS provides no built-in way to invoke a function asynchronously and poll for the invocation to finish. You can either call it synchronously, but that's potentially a 15 minute API call. Or invoke it asynchronously, but you have to invent your own mechanism for it to signal completion via another service. For now this implementation sticks with the former approach but it has some downsides, for example you need to make sure the HTTP connection doesn't time out.

romain-intel commented:
Thank you for this. I will try to take a look shortly.

savingoyal commented on extras/awslambda_image/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+RUN yum install -y vector wget
+
+RUN pip install awslambdaric metaflow awscli
```
Do we need to install `awslambdaric` if the lambda python image is used? Curious if there is a way to somehow not require `runtime interface client` in the image - that way any arbitrary image can be used for both AWS Batch & AWS Lambda : say the entry point installs `awslambdaric` first and then invokes `python -m awslambdaric`?

oavdeev replied:
You're right, we probably don't have to install it if we use the Lambda python image.

As for using an arbitrary image and installing `awslambdaric` on the fly, it [seems like](https://docs.aws.amazon.com/lambda/latest/dg/images-create.html#images-create-2) `awslambdaric` is not a pure Python package, so you'd have to have the gcc toolchain there as well. That seems like it could become pretty fragile. (I also wonder if AWS would kill the container prematurely if the runtime takes too long to initialize).

oavdeev replied:
Gotcha, I see, so the scenario you have in mind are users using Batch today and switching to Lambda?

One additional quirk to note here is that unlike e.g. Batch, your image for Lambda has to be a ECR image. Even [public ECR repo](https://docs.aws.amazon.com/AmazonECR/latest/public/public-repositories.html) doesn't cut it. 

So it seems like it would only benefit users that satisfy _all_ of the below conditions:
1) Use batch for these tasks today and want to move to `@awslambda`
2) ..and use custom images 
3) ..and store those images in ECR
4) ..and those images happen to have gcc toolchain to build awslambdaric

I'm not sure it is worth it, given that on-the-fly installation would be a bit hacky, and it would have to finish under undocumented deadline of what seems to be ~10s, ( [see this thread](https://stackoverflow.com/questions/55412624/does-aws-lambda-charge-for-the-time-spent-initializing-code/55426800) ) or possibly trigger more weird behaviors

oavdeev replied:
Discussed offline, we can't install awslambdaric on the fly as it has to be the entrypoint of the image - closing

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,213 @@
+)
+
+# Supported resource specs for Lambda
+LAMBDA_MAX_RESOURCES = {
```
I am not sure if hard coding these values (and runtime limit checks / timeouts) is a great idea since these are going to change with time as AWS releases new capabilities for AWS Lambda.

oavdeev replied:
They're mostly to have a nicer error message, and be able to show that error _before_ your flow is executed. Everything still works if we remove them. But i figured if e.g. the last step in a long flow has incorrect limits, it would be not so great UX to get an error only when you get there.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,213 @@
+        package: "MetaflowPackage",
+        run_id: str,
+    ):
+        if LAMBDA_LOG_GROUP_NAME is None:
```
Curious, what is the reasoning behind making `LOG_GROUP_NAME` user configurable?

oavdeev replied:
No huge reason, I imagine there could be people out there who have internal rules on naming the resources like CW group names.

savingoyal replied:
It might be good to rely on a default. Also, perhaps if the more recent comment in L:24 in `extras/awslambda_image/entrypoint.py` is amenable to you, we can avoid making the `LOG_GROUP_NAME` configurable at this moment. We haven't had any user requests so far for configuring the `LOG_GROUP_NAME` for AWS Batch executions.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,213 @@
+                "lambda role arn"
+            )
+
+        if LAMBDA_IMAGE_URI is None:
```
If no image is specified, we can default to a `python` image.

oavdeev replied:
unfortunately you can't use public images with Lambda, so the user would have to set this

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,213 @@
+        self.resource_attributes = {}
+        for deco in decos:
+            if isinstance(deco, CondaStepDecorator):
+                raise LambdaException(
```
What would it take to support `@conda` in `@awslambda`?

oavdeev replied:
One idea (mentioned it briefly in the PR decription) is to build custom docker images from conda envs, push them to ECR and use them to create Lambdas.

savingoyal replied:
Maybe, we can consider using [lambda layers](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html) rather than building images on the fly - which will presumably need more infrastructure. That can be a follow-on PR.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,213 @@
+    word in Python.
+    """
+
+    name = "awslambda"
```
thoughts on using `@lambda` instead of `@awslambda` to keep it consistent with `@batch`?

oavdeev replied:
That's what i had at first but `lambda` is a reserved word in Python

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,314 @@
+                        return
+
+
+def lambda_name(flow_name, step_name):
```
What if we constructed the `lambda_name` to be `flow_name-run_id-step_name-task_id-attempt_id` instead? That will also negate the need for introducing `force_fetch_image`, I believe?

oavdeev replied:
Yes it would, but it would add 5-30 seconds of extra latency to create the Lambda for each task. One of the goals of this PR was to add a way to run your code in the cloud with minimal latency so it feels delightful, so I was optimizing for lower startup time whenever I could here.

savingoyal replied:
We would have to necessarily isolate the `lambda` to at least the `run_id` which would mean that there is no specific need for `force_fetch_image`.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,314 @@
+    # satisfy the max of both requirements.
+    #
+    # [1] https://docs.aws.amazon.com/lambda/latest/dg/configuration-memory.html
+    if cpu is not None and memory_mb is not None:
```
This calculation runs the risk of becoming stale as AWS introduces new resource instances for Lambda. Is there a need for us to do this validation? Given that launching an AWS Lambda function is quick and easy - we can simply launch the function with the user-defined resources and return a properly formatted error if the values are incorrect.

oavdeev replied:
Re: validation, see the other comment. 

As for this math, lambda doesn't allow you to specify vCPUs directly, only memory, and you get X cpu for Y memory. I figured it would be nice if we still allowed users to specify CPU using `@resources` and did the math for them. The alternative is that we could say that `@resources` deco only supports specifying memory size for lambda, and we can remove this math. But users would have to do it anyway, and it is not trivial to dig it out of AWS docs.

savingoyal replied:
I think we can simply allow our users to only specify `memory`. As AWS changes the math on its end, it _might_ be tricky for us to guarantee backward compatibility.

oavdeev replied:
removed the cpu->memory_size math in c799b10

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,314 @@
+    # retry on conflict, as these updates should be idempotent anyway.
+    num_retries_on_conflict = 3
+
+    for retry in range(num_retries_on_conflict):
```
Will this logic work when I have two concurrent executions of a flow where the flow structure is the same but the resource specification is different for the same step?

oavdeev replied:
Nope, you're right, that would lead to weirdness. Those two executions would concurrently try to change Lambda resources and one of them would likely win

savingoyal replied:
Yeah, we should make sure we properly handle such scenarios.

oavdeev replied:
fixed in the last version

savingoyal commented on metaflow/plugins/aws/sync_metadata.py:
```diff
@@ -0,0 +1,62 @@
+import os
```
Minor nit: can we rename the file to `plugins/aws/utils.py`

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,314 @@
+from threading import Thread
+import boto3
+import json
+from typing import Dict, Optional, TYPE_CHECKING, Callable
```
It would be great to find a way to make AWS Lambda compatible with Py2 as well if it's not too much work.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,202 @@
+@click.argument("code-package-url")
+@click.option("--cpu", type=float, help="CPU requirement for AWS Lambda.")
+@click.option("--memory", type=int, help="Memory requirement for AWS Lambda.")
+@click.option(
```
Can we set this option as a global environment variable in `metaflow_config` instead? I presume the value of `throttle_retries` will stay the same across all executions for a user?

oavdeev replied:
done in c799b10

savingoyal reviewed (changes requested):
@oavdeev Great work! I took a quick look at the PR and noted some comments. Please take a look and let me know how you feel about any of those. I am yet to test the PR - I will follow up with more comments when I get to that 😃

oavdeev commented on extras/awslambda_image/entrypoint.py:
```diff
@@ -0,0 +1,88 @@
+        region = os.environ["AWS_REGION"]
+
+        # Create a vector config to pipe logs to Cloudwatch
+        with open(vector_config, "wt") as f:
```
They would, but finding them there is tricky. That CW stream name is determined by Lambda instance id that you don't control (instance = concrete running container), and multiple lambda invocations from the same instance go to the same stream. AFAIK AWS exposes no way to find the instance id corresponding to your lambda invoke() call.

In other words, the logs are indeed there, you only reliably know the Log Group name, and finding the logs for the concrete invocation would require multiple API calls (basically scanning all log streams).

The idea with vector here is to instead stream the logs to a CW log stream with a known name, separate for each task.

Anyway, `vector` requirement will go away once we merge that mflog PR

oavdeev commented on extras/awslambda_image/Dockerfile:
```diff
@@ -0,0 +1,12 @@
+
+RUN yum install -y vector wget
+
+RUN pip install awslambdaric metaflow awscli
```
done in c799b10

savingoyal commented on extras/awslambda_image/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM public.ecr.aws/lambda/python:3.7
```
We can move `extras` to Metaflow docs and metaflow-tools.

savingoyal commented on metaflow/metaflow_config.py:
```diff
@@ -125,6 +125,13 @@ def from_conf(name, default=None):
+###
+LAMBDA_ROLE_ARN = from_conf("METAFLOW_LAMBDA_ROLE_ARN")
+LAMBDA_IMAGE_URI = from_conf("METAFLOW_LAMBDA_IMAGE_URI")
+LAMBDA_THROTTLE_RETRIES = int(from_conf("METAFLOW_LAMBDA_THROTTLE_RETRIES", 20))
```
What is the utility of this variable? I see that it's used to specify `max-attempts` for the lambda invocation. We should rely on the value obtained from `@retry` instead.

oavdeev replied:
This parameter controls retries in case we get throttled by AWS due to Lambda parallel execution limits. I assume users will rarely need to touch this, but it is still nice to have an option. `@retry` is still respected and controls retries in case of step failure.

savingoyal replied:
Ah got it! Can we make a note in the code base about this behavior so that it doesn't throw off future committers?

savingoyal commented on metaflow/mflog/__init__.py:
```diff
@@ -1,4 +1,11 @@
+
+
+if (sys.version_info > (3, 0)):
+    from typing import Callable
```
Given that we are not using types at the moment, can we dispense with this import?

oavdeev replied:
The types are used below in `delayed_update_while`. Even though it is using the py2-compatible comment syntax (instead of py3 typing annotations), the type checker still needs the import to work. Really helps if you e.g. use vscode and have type checking enabled.

oavdeev replied:
addressed in #504

savingoyal commented on metaflow/mflog/mflog.py:
```diff
@@ -113,11 +113,9 @@ def refine(line, prefix=None, suffix=None):
-        return b''.join((header, b']', prefix, body, suffix))
-    else:
-        return line
+    assert len(parts) == 2
```
I am not a big fan of asserts. We can probably leave this segment of the code as-is for this PR.

oavdeev replied:
That's the bit we talked about offline, it seems like this is always called with mflog-formatted line so that `else` branch doesn't get called anyway. I can remove the assert itself given that it will fail on the next line anyway tho

oavdeev replied:
addressed in #504

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,204 @@
+@click.option("--run-id", help="Passed to the top-level 'step'.")
+@click.option("--task-id", help="Passed to the top-level 'step'.")
+@click.option("--image", help="Lambda image")
+@click.option(
```
I don't think `force_fetch_image` is needed any longer.

oavdeev replied:
Discussed offline -- will leave it in but as an advanced feature that we won't advertise. Normally users are expected to build an image under a new tag when it changes

oavdeev replied:
..and rename to `refresh_image`

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,204 @@
+        )
+
+    if R.use_r():
+        entrypoint = R.entrypoint()
```
To support R, we would need to make corresponding changes in `metaflow/R`

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,204 @@
+    input_paths = kwargs.get("input_paths")
+    split_vars = None
+    if input_paths:
+        max_size = 30 * 1024
```
Do we have the same limits in AWS Lambda as we have in AWS Batch?

oavdeev replied:
no good point i should remove this

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,204 @@
+            "max_attempts": LAMBDA_THROTTLE_RETRIES,
+        }
+    )
+    lambda_client = boto3.client("lambda", config=config)
```
We should use [`get_aws_client`](https://github.com/Netflix/metaflow/blob/002e7fd5c3be196e95222dce1972020f1f41d192/metaflow/plugins/aws/aws_client.py#L4) here

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_cli.py:
```diff
@@ -0,0 +1,204 @@
+        force_fetch_image=force_fetch_image,
+        echo=echo,
+    )
+    runner = LambdaRunner(
```
Given that the `lambda_client` is only used by the `LambdaRunner` - would it make sense to instantiate it within the `LambdaRunner` itself?

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+    LAMBDA_ROLE_ARN,
+)
+
+DEFAULT_TIMEOUT_SECONDS = 600
```
This is already available from https://github.com/Netflix/metaflow/pull/463/files#diff-cfe17a874aba5a08a3ed0ae7a9d22caa5eb63caa96bf392481edbdee323fc073R68

savingoyal replied:
Why is the default time out set to 10 mins when lambdas run for up to 15 mins?

oavdeev replied:
No strong reason. The default timeout on AWS is 3 seconds but that's too short for us (as we need to download the code too). 

On the other hand, if I try to put some really lengthy computation into a Lambda step, it would arguably be better UX if it failed faster and made me think about my choices and Lambda limitations. One could argue that if 10 minutes is not enough for your code, it is likely that 15 wouldn't be enough either.

So 10 minutes is just a point I picked between 3s and 10m. Maybe it should be even shorter?

savingoyal replied:
hmm.. If I am executing a flow on Lambda and Metaflow kills the job at 10m interval when I know that Lambdas run till 15 mins - that would be a confusing behavior. The only reason for having a timeout is to prevent long-running expensive jobs (AWS Batch timeout is 5 days for that very reason).

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+    name = "awslambda"
+    defaults = {
+        "image": None,
+        "force_fetch_image": False,
```
We can dispense with `force_fetch_image` - there is no particularly great way to handle this (I believe you want to handle the latest tag of the image here) and this issue is already present in AWS Batch/K8S etc. as well.

oavdeev replied:
Yes, this is mostly for the `:latest`. I'd have something similar for k8s and its `imagepullpolicy`, not sure how to handle AWS Batch tho.

Note that with Batch the problem is less acute, since you can hope that _eventually_ the new image will be used. With lambda, the image is copied by AWS when you create the function. So without this, there is just no way to trigger it to refetch the image, you'll be stuck with the old image forever. It was very annoying even when I was testing this on toy examples.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+
+            if isinstance(deco, BatchDecorator):
+                raise LambdaException(
+                    "You can't use @batch and *@awslambda* at the same time"
```
We need to move this check outside this routine and check for the validity of the flow before the flow starts executing.

oavdeev replied:
I agree but what's a good place for this? At first I looked at `FlowLinter` but it doesn't support plugins today

savingoyal replied:
I will commit a `PluginLinter` to the codebase to help with this.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+                        if k != "memory":
+                            raise LambdaException(
+                                "*@awslambda* only supports provisioning memory via @resources, not {r}".format(
+                                    r=k
```
I think rather than throwing an error, we can just ignore `cpu` etc. Makes it easier for folks to float between batch and lambda that way. Thoughts?

oavdeev replied:
Hmm good point, but it may be confusing for some if CPU and (especially) GPU settings are silently ignored but memory is not.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+                                )
+                            )
+                        else:
+                            if float(v) > LAMBDA_MAX_MEMORY_MB:
```
Given that the job will fail quickly if the user asks for more than the allowed memory limit, we can get rid of this check as well. As AWS raises the limits, we wouldn't have to upgrade our logic.

oavdeev replied:
I didn't have it at first but then I realized that doesn't fail _that_ quickly, it is only when it gets to that step you get an error. And it is a big ugly stacktrace. I think it makes a not great UX, and AWS only updated these limits twice in history, so in this I think it is worth it to put it on us to update the limits when it happens again. I'd feel differently if it was something that changes more frequently.

I understand that we probably can't do the same for each compute platform. IIRC it is much worse on Batch or k8s where you might get the job randomly stuck due to lack of resources and not get an error at all.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,208 @@
+                "The timeout for step *{step}* should be at "
+                "least 10 seconds for execution on AWS Lambda".format(step=step)
+            )
+        if self.run_time_limit > 900:
```
Ditto as the above comment.

oavdeev replied:
k will ignore it if the limit it too high to be consistent

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,291 @@
+
+    for retry in range(num_retries_on_conflict):
+        try:
+            return _maybe_update_lambda(
```
I thought we don't need this code segment anymore?

oavdeev replied:
We do since there may be a race condition where multiple branches of foreach try to create the same lambda simultaneously. It is much simpler than the previous version though.

savingoyal replied:
Is this still needed after the move to `package_init`?

savingoyal reviewed (changes requested):
Still reviewing - here are some partial comments.

oavdeev commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,291 @@
+            )
+        except client.exceptions.ResourceConflictException:
+            if retry == num_retries_on_conflict - 1:
+                print("ENOUGH CONFLICT RETRIES")
```
forgot to remove debug print here

oavdeev commented:
Rebased this on #504 and addressed PR comments, except two things:
* i can't find a better place to create lambdas than `task_pre_step`. Tried moving it to `step_init` but that gets called _inside_ Lambda too which is not what I want. I could probably work around this but feels like it would unnecessarily complicate logic
* prefixing logs with request id turned out to be suuper cumbersome due to the fact that request id is only available inside the lambda during its execution, so I would have to do it in a hacky way that is pretty different how prefixing is done for `@batch`. So instead I made it prefix logs with just the lambda name. Request id is not super useful anyway.

oavdeev commented:
@savingoyal made it create lambdas in `package_init`

savingoyal commented on metaflow/util.py:
```diff
@@ -129,7 +129,7 @@ def to_unicode(x):
         return unicode_type(x)
```
Minor nit: Can we remove this file from the PR?

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,264 @@
+
+        # Run time limit more than 15m is effectively ignored
+        self.run_time_limit = min(
+            get_run_time_limit_for_task(decos, default_limit=DEFAULT_TIMEOUT_SECONDS),
```
Curious what happens if we leave the run-time limit to 5 days? Does AWS Lambda complain?

savingoyal replied:
Also, when does the run-time-limit clock start?

oavdeev replied:
Yeah you can't set it above 900 seconds. Starts when Lambda is invoked.

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_decorator.py:
```diff
@@ -0,0 +1,264 @@
+        meta["aws-lambda-execution-env"] = os.environ["AWS_EXECUTION_ENV"]
+
+        # Unlike the above vars, this is set by our own code in entrypoint.py
+        meta["aws-lambda-request-id"] = os.environ.get("LAMBDA_REQUEST_ID", "")
```
Since this is guaranteed to exist - should we just do `os.environ["LAMBDA_REQUEST_ID"]`

savingoyal commented on metaflow/plugins/aws/awslambda/lambda_runner.py:
```diff
@@ -0,0 +1,293 @@
+
+        # If refresh_image is set, force update function code.
+        if refresh_image:
+            echo("Updating image for %s.." % name, False)
```
Given that this is an internal implementation detail - should we print this? Same comment for L:178.

oavdeev replied:
I kinda went back and forth on this; the main reason for having this is that it can often take about 30 seconds to update or create lambda (if it doesn't exist or `refresh_image` is used). So I wanted to give the user some indication that something is happening, Metaflow isn't magically "stuck".

savingoyal replied:
We can have a different message to address that use case. For example - AWS Batch cycles through the batch status codes so that the user understands that the job is in progress - we print a message when the status changes or if it's been 30s since the last message. We can have a similar behavior here.

savingoyal replied:
Can we have the same(similar) message as we have for AWS Batch?

oavdeev replied:
Hmm the batch is a different beast so the message is different, there is no "job status" in the case of Lambda. And note that is not a direct equivalent of what Batch is doing, you'll only see this message on the first run when Lambda resource needs to be created.

So the copy here is the best I could come up with -- but again, open to concrete alternative suggestions

savingoyal reviewed (commented):
I had a few pending comments that I forgot to publish.

romain-intel commented on metaflow/decorators.py:
```diff
@@ -208,7 +208,7 @@ def step_init(self, flow, graph, step_name, decorators, environment, datastore,
         pass
 
-    def package_init(self, flow, step_name, environment):
+    def package_init(self, flow, step_name, environment, logger):
```
Not sure this change is needed here. You can use the logger from step_init IIRC.

romain-intel reviewed (commented):
Didn't look in detail at the AWS lambda aspect (I don't know that very well). A few general comments though:
  - the mflog related changes should ideally be extracted to a separate PR (they are not related to AWS per say and it would be nice to have those singled out)
  - you can consider also stripping out the refactor on batch if you want (though that one is less of an issue)
  - if we are to merge this pre convergence branch, please also make the same PR(s) available on the convergence branch; the datastore changes will have an impact on this (no functional difference). We are going to merge back master into convergence soon so maybe after that, it would be good to make a parallel PR chain against that one as well.
 
I will run some tests as well (just quickly looked at it to give a quick initial feedback).

oavdeev commented:
@romain-intel the mflog changes are already factored out to https://github.com/Netflix/metaflow/pull/504 . Unfortunately Github still doesn't make it easy to manage "stacked" PRs, so this PR branch has mflog PR merged in, and those changes appear in the diff view :/

romain-intel commented:
Ah ok. Yes, stacked PRs are a pain (boy do I know that). You can do it manually but not as much fun. OK. I'll look at those independently as well and make sure everything still works internally.

oavdeev commented:
@gatesyp we're pretty close! Just had to pause before some refactorings work for data store gets merged in, which should happen next week. Then the plan is to rebase this, fix some nits and merge

gatesyp commented:
Ok! Sounds great :) if I'm understanding the PR correctly, this would be pretty important to my group. If there's anything I can do let me know! More than willing to help out

oavdeev commented:
We're focusing our efforts more on K8S support right now so unfortunately we don't have bandwidth to drive this Lambda PR to conclusion. More importantly, we want to make sure that every feature that is merged into Metaflow core is well supported going forward, e.g. works well with all future orchestrators we may introduce. Realistically we can't commit to this for AWS Lambda even if we merge the PR as is today. 

So I'm going to close this PR since its not actively being worked on.

It is possible that we'll revisit this decision in the future, so feel free to chime in if you really want Lambda or talk to us on Slack.
